### DISCUSSION TITLE
Sending .wav file to streaming api in nodejs?

### DISCUSSION THREAD

**Discussion by jdheeter** [deepgram/community#523] [General help] [answered]
I'm working on a project that uses the live streaming endpoint to stream microphone data to deepgram and it works great. For testing sometimes I need to send pre-recorded data in .wav along the same socket connection. I tried just reading the data and sending the Buffer object on the connection using .send() and I get no response back on the socket. I was expecting at least an error/warning to let me know what is wrong.

I found this python example that can send .wav on the streaming API, but not sure how to translate this to nodejs.
https://github.com/deepgram/streaming-test-suite/blob/main/server.py

Yes I know the streaming API is not designed for static files like this but I'm just using this for repeatable testing (trying to test latency in our system) so I'd like to use streaming API for this test to match the live setup.

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jdheeter** [comment]
So ChatGPT was able to help me convert the example python client to nodejs code and this works.

```ts
import * as fs from "fs"
import WebSocket from "ws"
import { Buffer } from "buffer"
import env from "lib/env"
import { log } from "console"

const REALTIME_RESOLUTION = 0.250
const encodingSampleWidthMap:{ [key:string]:number } = {
  "linear16": 2,
  "mulaw": 1
}

async function audioStream(audioFilePath:string, encoding:string, sampleRate:number, channels:number):Promise<void> {
  const data = fs.readFileSync(audioFilePath)
  let buffer = Buffer.from(data)

  const url = `wss://api.deepgram.com/v1/listen?encoding=${encoding}&sample_rate=${sampleRate}&channels=${channels}?model=nova-2-general`
  const ws = new WebSocket(url, {
    headers: {
      "Authorization": "Token " + env.DEEPGRAM_KEY
    }
  })

  ws.on("open", () => {
    console.log("🟢 (1/5) Successfully opened streaming connection")
    void sender(ws, buffer, encoding, sampleRate, channels)
  })

  ws.on("message", (msg) => {
    console.log("🟢 (3/5) Successfully receiving server messages")
    const res = JSON.parse(msg.toString())
    log(res)
    log(res.channel)
    // Handling of the response goes here
  })
}

async function sender(ws:WebSocket, data:Buffer, encoding:string, sampleRate:number, channels:number):Promise<void> {
  console.log("🟢 (2/5) Ready to stream data")

  const sampleWidth = encodingSampleWidthMap[encoding]
  const byteRate = sampleWidth ? sampleWidth * sampleRate * channels : undefined
  const chunkSize = byteRate ? Math.floor(byteRate * REALTIME_RESOLUTION) : 5000

  while (data.length > 0) {
    const chunk = data.slice(0, chunkSize)
    data = data.slice(chunkSize)
    await new Promise(resolve => setTimeout(resolve, REALTIME_RESOLUTION * 1000))
    ws.send(chunk)
  }

  ws.send(JSON.stringify({ type: "CloseStream" }))
  console.log("🟢 (4/5) Successfully closed connection, waiting for final messages if necessary")
}

// Example usage
// await audioStream("../preamble.wav", "linear16", 44100, 2)
await audioStream("../audioScripts/test1/1.wav", "linear16", 48000, 1)
```

however when I try to use the library to achieve the same result
```
 this.deepgramEvents = deepgram.listen.live({
            model: "nova-2-general",
            encoding: "linear16",
            channels: 1,
            sample_rate: 48000
          })
```
I get no response from the websocket when I send the data using this function
```ts
async function sendDGAudioFile(ws:LiveClient, data:Buffer, encoding:string, sampleRate:number, channels:number):Promise<void> {
  const REALTIME_RESOLUTION = 0.250
  const encodingSampleWidthMap:{ [key:string]:number } = {
    "linear16": 2,
    "mulaw": 1
  }

  const sampleWidth = encodingSampleWidthMap[encoding]
  const byteRate = sampleWidth ? sampleWidth * sampleRate * channels : undefined
  const chunkSize = byteRate ? Math.floor(byteRate * REALTIME_RESOLUTION) : 5000

  while (data.length > 0) {
    const chunk = data.slice(0, chunkSize)
    data = data.slice(chunkSize)
    await new Promise(resolve => setTimeout(resolve, REALTIME_RESOLUTION * 1000))
    log.debug("Sending chunk, size: ", chunk.length)
    ws.send(chunk)
  }
}
```
```ts
sendDGAudioFile(this.deepgramEvents, audioData, "linear16", 48000, 1).catch(console.error)
```

I checked the generated `LiveClient` object ._socket.url parameter is `wss://api.deepgram.com/v1/listen?encoding=linear16&sample_rate=48000&channels=1?model=nova-2-general` Which seems correct. When I paste this url into the converted example code at the top of this post, I get the correct response from the API that I expect. So it seems like there is something different about the way the api client sends data that I'm not understanding.

Appreciate any pointers!

**Comment by jdheeter** [comment]
I found the reason my code was not receiving any result. It's because the example code calls finish() as soon as the file chunks are finished sending. I'm not sure if it's possible to send multiple files along the same socket connection? It seems like you have to call .finish() to get any result at all. It would be nice if you could finish but keep the connection over for more files.

**Reply by lukeocodes** [reply, marked as answer]
I think you may be better off testing what you actually want to build. Sending .wav files down to the live websocket might technically be possible, but it is complicated for the sake of testing out Deepgram.

We have a [Node Live audio Starter app](https://github.com/deepgram-starters/live-node-starter) you might want to check out. This probably does a lot of what you want to ultimately do.

**Reply by jdheeter** [reply]
I have the code working streaming audio from my mic I just wanted to be able to test the latency in a repeatable way by sending a .wav file on the same socket connection.

**Reply by jdheeter** [reply]
It would be good if the Node Starter Kit you linked above was easily visible in the docs. The Live Streaming Starter Kit page only references the python version

https://developers.deepgram.com/docs/getting-started-with-the-streaming-test-suite
